awx/main/models/credential/__init__.py

```diff
@@ -461,8 +472,18 @@ def get_input(self, field_name, **kwargs):
         raise AttributeError(field_name)
 
     def has_input(self, field_name):
+        if field_name in self.dynamic_input_fields:
+            return True
         return field_name in self.inputs and self.inputs[field_name] not in ('', None)
 
+    def _get_dynamic_input_field_names(self):
+        input_sources = CredentialInputSource.objects.filter(target_credential=self)
```

Comment:
I think it's better to use the related name when possible, in this case, `input_source` I believe (it might be good to make this plural). That will use the prefetched objects if someone does `Credential.objects.prefetch_related('input_source')`. It might still use it in this case, but I think related name is better. That said, it still shouldn't be called from a model's `__init__`.

Reply:
Agreed on using the related names when possible. I _think_ the reason I bypassed them here was to get around some credential `__getattr__` behavior but I'll double-check. If the related names work we can definitely switch to using those.

Reply:
It's possible that the entirety of this method (and the other one) could be reduced to

```python
    @cached_property
    def dynamic_input_fields(self):
        return [obj.input_field_name for obj in self.input_sources.all()]
```